### PR TITLE
Move obvious type piracy from Hecke to Nemo

### PR DESCRIPTION
**Pull request by lgoettgens**
Resolves parts of https://github.com/thofma/Hecke.jl/issues/1126.

I will have a look through the Hecke files that contain many cases of reported type piracy, and move those functions to Nemo, that are either trivial in their implementation, or if it is clear that they only use functionality already in Nemo (like calls to flint).

Functions, where I do not directly see a good place to put them, will be in some files named after the Hecke file they originate from for the time being. Moving them around in Nemo is easy to do later in a separate PR. Or if you have any recommendations, let me know and I will move them in this PR.

The then duplicate functions will be removed from Hecke at https://github.com/thofma/Hecke.jl/pull/1143.

### CODE REVIEW

File: src/HeckeMiscInteger.jl [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+function isless(a::BigFloat, b::ZZRingElem)
+    if _fmpz_is_small(b)
+        c = ccall((:mpfr_cmp_si, :libmpfr), Int32, (Ref{BigFloat}, Int), a, b.d)
+    else
+        c = ccall((:mpfr_cmp_z, :libmpfr), Int32, (Ref{BigFloat}, UInt), a, unsigned(b.d) << 2)
+    end
+    return c < 0
+end
+
+function mulmod(a::UInt, b::UInt, n::UInt, ni::UInt)
```
Comment (fingolfin) [inline src/HeckeMiscInteger.jl:20]: There are probably better places for this and other functions. What's the plan here? Move the first into this file and merge it; and then when we are done, we can take our time to move things to a nicer place?

**Comment by fingolfin** [comment]
Maybe it would be best to organize a meeting (real or virtual) involving at least @lgoettgens @thofma and @fieker (I am also happy to join) to closely coordinate these changes and do them in 1-2 days, instead of having all these disruptive changes spaced out over multiple weeks and PRs and so on....

File: src/HeckeMiscInteger.jl [hunk 2]
```diff
@@ -0,0 +1,438 @@
+################################################################################
+#
+#  Integer functions
+#
+################################################################################
+
+function rem(a::ZZRingElem, b::UInt)
+    return ccall((:fmpz_fdiv_ui, libflint), UInt, (Ref{ZZRingElem}, UInt), a, b)
+end
```
Comment (fingolfin) [inline src/HeckeMiscInteger.jl:9]: `src/flint/fmpz.jl`

File: src/HeckeMiscInteger.jl (hunk 1 above)
Comment (fingolfin) [reply src/HeckeMiscInteger.jl:20]: `src/flint/fmpz.jl`?

File: src/HeckeMiscInteger.jl [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    end
+    if a == -1
+        if isodd(k)
+            return QQFieldElem(-1)
+        else
+            return QQFieldElem(1)
+        end
+    end
+    return a^Int(k)
+end
```
Comment (fingolfin) [inline src/HeckeMiscInteger.jl:156]: `src/flint/fmpq.jl`

File: src/HeckeMiscInteger.jl [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+function //(a::ZZRingElem, b::QQFieldElem)
+    return QQFieldElem(a) // b
+end
```
Comment (fingolfin) [inline src/HeckeMiscInteger.jl:160]: `src/flint/fmpq.jl`

File: src/HeckeMiscInteger.jl [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+log(a::ZZRingElem) = log(BigInt(a))
+log(a::QQFieldElem) = log(numerator(a)) - log(denominator(a))
+
+function log(a::ZZRingElem, b::ZZRingElem)
+    log(b) / log(a)
+end
+
+function divisible(x::Integer, y::Integer)
+    return iszero(rem(x, y))
+end
```
Comment (fingolfin) [inline src/HeckeMiscInteger.jl:199]: for AbstractAlgebra?

**Review by fingolfin — COMMENTED** [review]
A couple more semi-random suggestions. I'd like to stress that these are just meant as food for thought. I don't mind merging the PR w/o them (of course there are other blockers for merging, like tests need to pass and syncing with our other packages be sorted out).

File: src/HeckeMiscInteger.jl [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    end
+    return v
+end
+
+function valuation(z::Rational{T}, p::T) where {T<:Integer}
+    z == 0 && error("Not yet implemented")
+    v = valuation(denominator(z), p)
+    w = valuation(numerator(z), p)
+    return w - v
+end
```
[the remark this answers is not in the document]
Comment (lgoettgens) [reply src/HeckeMiscInteger.jl]: I do not know if it is easier to manage large PRs over three repos right now or first move stuff from Hecke to Nemo and then later, in a separate PR, tackle the Nemo <-> AbstractAlgebra side. For the moment, I am doing the later.

**Comment by lgoettgens** [comment]
The number of reported piracy cases already went down from 721 to 425!

File: src/HeckeMiscMatrix.jl [hunk 7]
```diff
@@ -0,0 +1,1225 @@
...
+
+################################################################################
+#
+#  Concatenation of matrices
+#
+################################################################################
+
+@doc raw"""
+vcat(A::Vector{Generic.Mat}) -> Generic.Mat
+vcat(A::Array{ZZMatrix}, 1}) -> ZZMatrix
```
Comment (fingolfin) [inline src/HeckeMiscMatrix.jl:658]: Seems my scripts missed this one when we changed Array -> Vector/Matrix :-(
```suggestion
vcat(A::Vector{ZZMatrix}) -> ZZMatrix
```

**Comment by lgoettgens** [comment]
The OscarCI-release failures are expected as this is a breaking change.
The OscarCI-matching-1.8 failures will be fixed by https://github.com/oscar-system/Oscar.jl/pull/2442.
As the OscarCI-matching-1.6 succeeded, I think this is good to go.

File: src/HeckeMiscInteger.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgoettgens) [reply src/HeckeMiscInteger.jl]: Changed my plan. There now is https://github.com/Nemocas/AbstractAlgebra.jl/pull/1392 containing the parts of this PR that belong to AA.

**Comment by lgoettgens** [comment]
The now relevant CI for everything working is the OscarCI-matching in https://github.com/Nemocas/AbstractAlgebra.jl/pull/1392.

**Comment by lgoettgens** [comment]
The move of some stuff to AbstractAlgebra seemed too involved for me to fix it in the next days. Thus I would revert back to the original plan to just move a lot of stuff from Hecke to Nemo in this PR, and look to Nemo and AA sometime in the future.
I guess we will have enough fun with the other breaking changes in AA for the next release.

**Pull request merged**